### PR TITLE
Simple replica allocator tests

### PR DESCRIPTION
**Pull request by cuongdo**
Added the following tests:

* 1->3 nodes
* 3->5 nodes

Each test involves 10G of data / 262 ranges.

Note that this is not yet automated. It's meant to be run manually. That's a TODO for another PR in the near future.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach-prod/89)

### CODE REVIEW

**Comment by cuongdo** [comment]
cc @jordanlewis @WillHaack

**Comment by jordanlewis** [comment]
Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[allocatortest/allocatortest-common.sh, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89#-KKKqhSchpjwIu8GR7v3:-KKKqhSchpjwIu8GR7v4:-1774389431) ([raw file](https://github.com/cockroachdb/cockroach-prod/blob/b663937632998771530c24a81ee48d26014cf04d/allocatortest/allocatortest-common.sh#L36)):*
> ```Shell
>     which terraform >/dev/null 2>&1 || die "Couldn't find terraform"
>     which gcloud >/dev/null 2>&1 || die "Couldn't find gcloud. Is the Google Cloud SDK installed?"
>     which gsutil >/dev/null 2>&1 || die "Couldn't find gcloud. Is the Google Cloud SDK installed?"
> ```

Small typo - probably meant gsutil here.

---

*[allocatortest/README.md, line 9 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89#-KKKr4QxEoG6I6Dxji3u:-KKKr4QxEoG6I6Dxji3v:533220740) ([raw file](https://github.com/cockroachdb/cockroach-prod/blob/b663937632998771530c24a81ee48d26014cf04d/allocatortest/README.md#L9)):*
> ```Markdown
> 1. Make sure you have a [Google Cloud Platform](https://cloud.google.com/compute/) account.
> 2. [Download terraform](https://terraform.io/downloads.html), *version 0.6.7 or greater*, unzip, and add to your `PATH`.
> 3. Install `dsh`.
> ```

you should include a section on installing gcloud and the other tool that's required above.

---

*[allocatortest/README.md, line 67 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89#-KKKrBVuuC54YRIgSXzf:-KKKrBVuuC54YRIgSXzg:-221918526) ([raw file](https://github.com/cockroachdb/cockroach-prod/blob/b663937632998771530c24a81ee48d26014cf04d/allocatortest/README.md#L67)):*
> ```Markdown
> 
> When run without any parameters, each test will execute from beginning to end. However, to take full control over the test cluster, you can pass the following arguments to the test:
> ```

I really like this stuff. I think we should consider moving some of these operations into a separate tool. I could see it being enormously useful for other tests besides this one, as well as people doing random administrative cluster tasks. cdb_util?

---

*[allocatortest/terraform/download_binary.sh, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89#-KKKrgns9poHZia4Hzt4:-KKKrgntIGVfhjuPJG0A:-1996015303) ([raw file](https://github.com/cockroachdb/cockroach-prod/blob/b663937632998771530c24a81ee48d26014cf04d/allocatortest/terraform/download_binary.sh#L3)):*
> ```Shell
> #!/bin/bash
> # This script takes the name of a binary and downloads it from S3.
> # It takes the repo_name/binary_name and an optional sha.
> ```

Doesn't this script exist somewhere else? I thought I recalled seeing it elsewhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89)*

**Comment by cuongdo** [comment]
cc @petermattis

File: allocatortest/allocatortest-common.sh [hunk 1]
```diff
@@ -0,0 +1,312 @@
+# Copyright 2016 The Cockroach Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http:#www.apache.org/licenses/LICENSE-2.0
```
Comment (petermattis) [inline allocatortest/allocatortest-common.sh]: `s,http:#,http://,g`

**Comment by petermattis** [comment]
Shell is not a maintainable language. I understand the desire to use it for these kinds of tests (I've done similar stuff myself), but I think we should quickly be moving to convert this test to Go and to enhance the `acceptance/terrafarm` infrastructure to accommodate tests like this and making developing and debugging such tests pleasant. Don't let this feedback hold up merging. Just my two cents.

**Comment by cuongdo** [comment]
I agree bash scripting is not ideal. I'll experiment with `terrafarm` and see if I can make it work.

The trick here is to come up with a framework that allows easy:
1. Creation of tests
2. Fixing of broken tests
3. Pause/Resume/Push new `cockroach` binaries of in-progress tests
4. Manual SSH into clusters to see what's going on (e.g. to get pprof profiles or `tail -f` logs)

When tests take a minimum of 3-5 minutes for initial Terraform setup (and another few minutes to destroy) and 20-40 minutes to run, anything that can be done to reduce iteration times is key.

One thing that would help is a "fast restart" test feature that given an existing test cluster, stops CockroachDB, wipes store data, repopulates data, and restarts CockroachDB without re-Terraforming.

**Comment by tschottdorf** [comment]
> (and another few minutes to destroy) 

In a Go-based test runner, you could conceivably let the destroy operations run and already start the next test. Just in case that ever turns into a sizable nuisance.

**Comment by jordanlewis** [comment]
I like the fast restart feature. I think that it might not be too hard to implement. Terrafarm could be aware of some per-user (and maybe per-test?) resource prefix that it checks existence of in GCE before spinning up new clusters. That way you could spin up a persistent test cluster for quicker iteration and terrafarm wouldn't bother creating or destroying resources.

Alternatively, we could simply add a new flag to terrafarm that points to an existing set of resources.

**Comment by petermattis** [comment]
Let's get this merged and use @cuongdo's requirements as a driver for enhancements to terrafarm. A short term goal should be getting this test converted to `acceptance/terrafarm`.

**Comment by bdarnell** [comment]
Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[allocatortest/allocatortest-common.sh, line 213 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89#-KKPqihrF4t95PMtXeOW:-KKPqihrF4t95PMtXeOX:538196005) ([raw file](https://github.com/cockroachdb/cockroach-prod/blob/b663937632998771530c24a81ee48d26014cf04d/allocatortest/allocatortest-common.sh#L213)):*
> ```Shell
>         echo "* Copying logs from ${ip} (node ${node_id})"
>         node_id=$((node_id+1))
>         (set -x; scp ${SCP_FLAGS} -Cpr ubuntu@${ip}:logs "${logs_dir}/${node_id}")
> ```

FYI there's a package called `pssh` that includes a multi-host version of `scp`. I still use `dsh` out of habit, but `pssh` is nice for downloading files. 

---

*[allocatortest/README.md, line 18 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89#-KKPrAEoRBeVNRtzdVML:-KKPrAEoRBeVNRtzdVMM:1137242453) ([raw file](https://github.com/cockroachdb/cockroach-prod/blob/b663937632998771530c24a81ee48d26014cf04d/allocatortest/README.md#L18)):*
> ```Markdown
> 6. Save your GCE key as `~/.ssh/google_compute_engine`.
> 7. Install `wait_rebalance`: `go install -v github.com/cockroachdb/cockroach-prod/tools/wait_rebalance`
> 8. Run `make clusterctl` in your top-level `cockroach` directory.
> ```

What's `clusterctl`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach-prod/89)*

**Comment by cuongdo** [comment]
Merging with the various imperfections, with the understanding that this will be ported to `terrafarm` soon

**Pull request merged**